pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -143,6 +151,48 @@ func (a *admitter) admitCommonCreateUpdate(oldProject, newProject *v3.Project) (
 	return admission.ResponseAllowed(), nil
 }
 
+func (a *admitter) validateContainerDefaultResourceLimit(limit *v3.ContainerResourceLimit) error {
+	if limit == nil {
+		return nil
+	}
+	fieldPath := projectSpecFieldPath.Child(containerLimitField)
+	requestsCPU, err := parseResource(limit.RequestsCPU)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.RequestsCPU, fmt.Sprintf("failed to parse container default requested CPU: %s", err))
+	}
+	limitCPU, err := parseResource(limit.LimitsCPU)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.LimitsCPU, fmt.Sprintf("failed to parse container default CPU limit: %s", err))
+	}
+	requestsMemory, err := parseResource(limit.RequestsMemory)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.RequestsMemory, fmt.Sprintf("failed to parse container default requested memory: %s", err))
+	}
+	limitMemory, err := parseResource(limit.LimitsMemory)
+	if err != nil {
+		return field.Invalid(fieldPath, limit.LimitsMemory, fmt.Sprintf("failed to parse container default memory limit: %s", err))
+	}
+	resourceList := v1.ResourceList{
+		v1.ResourceRequestsCPU:    requestsCPU,
+		v1.ResourceLimitsCPU:      limitCPU,
+		v1.ResourceRequestsMemory: requestsMemory,
+		v1.ResourceLimitsMemory:   limitMemory,
+	}
+	if negativeResources := quota.IsNegative(resourceList); len(negativeResources) > 0 {
+		return field.Invalid(fieldPath, limit,
+			fmt.Sprintf("resource requests and limits can't have negative values, adjust the request/limit configurations for the following resources: %v", negativeResources))
+	}
+	if !requestsCPU.IsZero() && !limitCPU.IsZero() && requestsCPU.Cmp(limitCPU) > 0 {
```

Comment:
Unfortunately I don't think that this works. When I try to parse a 0 value on my local it returns a quantity that evaluates to true for this statement:

```golang
zeroStringM, err := resource.ParseQuantity("0m")
if err != nil {
	panic("error when parsing 0m string")
}
// prints 0m string is zero: true 
fmt.Printf("0m string is zero: %t \n", zeroStringM.IsZero())
```

This means that the following limit range would pass your webhook check (since the CPU limit is zero):
```yaml
apiVersion: v1
kind: LimitRange
metadata:
  name: test
spec:
  limits:
  - default:
      cpu: 0m
    defaultRequest:
      cpu: 1000m
    type: Container
```

However, when I try to create this in k8s, I get an error:
```
The LimitRange "test" is invalid: spec.limits[0].defaultRequest[cpu]: Invalid value: resource.Quantity{i:resource.int64Amount{value:1000, scale:-3}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"1", Format:"DecimalSI"}: default request value 1 is greater than default limit value 0
```

I think that this will cause the original defect to present under certain circumstances (namely the use of a 0 quantity) and would recommend fixing this.

Reply:
Good catch! I added those test cases and changed the logic, so that it's correct and also doesn't use the "zero" terms, which are ambiguous.